cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/DefaultSecureStoreService.java

```diff
@@ -0,0 +1,174 @@
+import co.cask.cdap.proto.id.SecureKeyId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.proto.security.SecureKeyCreateRequest;
+import co.cask.cdap.proto.security.SecureKeyListEntry;
+import co.cask.cdap.security.authorization.AuthorizerInstantiator;
+import co.cask.cdap.security.spi.authentication.AuthenticationContext;
+import co.cask.cdap.security.spi.authentication.SecurityRequestContext;
+import co.cask.cdap.security.spi.authorization.AuthorizationEnforcer;
+import co.cask.cdap.security.spi.authorization.Authorizer;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Default implementation of the service that manages access to the Secure Store,
+ */
+public class DefaultSecureStoreService implements SecureStoreService {
+  private final Authorizer authorizer;
+  private final AuthorizationEnforcer authorizationEnforcer;
+  private final AuthenticationContext authenticationContext;
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+  @Inject
+  DefaultSecureStoreService(AuthorizerInstantiator authorizerInstantiator, AuthorizationEnforcer authorizationEnforcer,
+                            AuthenticationContext authenticationContext, SecureStore secureStore,
+                            SecureStoreManager secureStoreManager) {
+    this.authorizer = authorizerInstantiator.get();
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.authenticationContext = authenticationContext;
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+  }
+
+  /**
+   * Lists all the secure keys in the given namespace that the user has access to.
+   * @param namespaceId Id of the namespace we want the key list for.
+   * @return A list of {@link SecureKeyListEntry} of all the keys visible to the user under the given namespace.
+   * @throws IOException If there was a problem getting the list from the underlying provider.
+   * @throws UnauthorizedException If we fail to create a filter for the current principal.
+   *
+   */
+  @Override
+  public List<SecureKeyListEntry> list(NamespaceId namespaceId) throws IOException, UnauthorizedException {
+    Principal principal = authenticationContext.getPrincipal();
+    List<SecureStoreMetadata> metadatas = secureStore.list(namespaceId.getNamespace());
+    List<SecureKeyListEntry> returnList = new ArrayList<>(metadatas.size());
+    final Predicate<EntityId> filter;
+    try {
+      filter = authorizationEnforcer.createFilter(principal);
+    } catch (Exception e) {
+      throw new UnauthorizedException(principal, Action.READ, namespaceId);
```

Comment:
technically this could fail for some other reason right?

Also, list should not throw an exception. it should just filter